ProfessorAssessor/View Controllers/ProfessorViewController.m

```diff
@@ -39,35 +52,189 @@ - (void)setProfessorDetails {
-    self.selectedCourses = [NSMutableArray new];
+    self.selectedCourses = [NSMutableSet new];
 
     [self displayCourseTags];
 
-    [self fetchReviews];
+    [self fetchReviewsForTagChange];
+}
+
+- (void)fetchReviewsWithLimit:(int)limit
+               withCompletion:(void (^)(NSArray<Review *> *_Nullable objects,
+                                        NSError *_Nullable error))completion {
+    self.query = [Networker fetchReviewsForProfessor:self.professor
+                                          forCourses:self.selectedCourses
+                                               limit:limit
+                                      withCompletion:completion];
 }
 
-- (void)fetchReviews {
+- (void)fetchReviewsForRefresh {
+    if (self.query) {
+        [self cancelQuery];
+    }
+
+    [self.activityIndicator startAnimating];
+
     __weak typeof(self) weakSelf = self;
 
-    [Networker
-     fetchReviewsForProfessor:weakSelf.professor
-     forCourses:weakSelf.selectedCourses
-     withCompletion:^(NSArray<Review *> *_Nullable objects,
-                      NSError *_Nullable error) {
-        if (objects) {
+    [self fetchReviewsWithLimit:self.queryLimit
+                 withCompletion:^(NSArray<Review *> *_Nullable objects,
+                                  NSError *_Nullable error) {
+        __strong __typeof(self) strongSelf = weakSelf;
+
+        strongSelf.reviews = objects;
+
+        [strongSelf.tableView reloadData];
+        [strongSelf.tableView.refreshControl endRefreshing];
+        [strongSelf.activityIndicator stopAnimating];
+
+        strongSelf.averageRating.value = [strongSelf.professor.averageRating doubleValue];
+
+        strongSelf.query = nil;
+    }];
+}
+
+- (void)fetchReviewsForInfiniteScroll {
+    if (self.query) {
+        [self cancelQuery];
+    }
+
+    NSArray<Review *> *reviews = [self.reviewCache objectForKey:self.selectedCourses];
+
+    if (reviews.count >= self.queryLimit) {
+        self.reviews = reviews;
+
+        [self.tableView reloadData];
+        [self.loadingMoreView stopAnimating];
+
+        self.queryLimit = (int)reviews.count + queryLimitIncrement;
+        self.loadingMoreReviews = NO;
+    } else {
+        __weak typeof(self) weakSelf = self;
+        NSMutableSet<Course *> *selectedCourses = [self.selectedCourses copy];
+
+        [self fetchReviewsWithLimit:self.queryLimit
+                     withCompletion:^(NSArray<Review *> *_Nullable objects,
+                                      NSError *_Nullable error) {
+            __strong __typeof(self) strongSelf = weakSelf;
+            NSArray<Review *> *reviews = [objects mutableCopy];
+
+            strongSelf.reviews = objects;
+
+            [strongSelf.reviewCache setObject:reviews forKey:selectedCourses];
+
+            [strongSelf.tableView reloadData];
+            [strongSelf.loadingMoreView stopAnimating];
+
+            strongSelf.loadingMoreReviews = NO;
+
+            strongSelf.query = nil;
```

Comment:
Not a huge deal but I'll often break completions into methods so that it's easier to read. Then you just call, "self.completeSomething(with: thing-you-go-back)"